### PR TITLE
feat deploy service with the same multisig using multisend

### PR DESCRIPTION
**Pull request by kupermind**
- A concept of giving all the rights of the terminated service multisig to the service owner;
- The service is then updated;
- The new agent instances are added to the multisig by the service / multisig owner;
- The new service is deployed with the same multisig.

### CODE REVIEW

File: test/ServiceRegistry.js [hunk 1]
```diff
@@ -905,6 +905,111 @@ describe("ServiceRegistry", function () {
...
+            const safeContracts = require("@gnosis.pm/safe-contracts");
+            // At this point of time the agent instance gives the ownership rights to the service owner
+            // In other words, swap the owner of the multisig to the service owner (agent instance to give up rights for the service owner)
+            // Since there was only one agent instance, the previous multisig owner address is the sentinel one defined by gnosis (0x1)
+            const sentinelOwners = "0x" + "0".repeat(39) + "1";
+            let nonce = await multisig.nonce();
+            const txHashData = await safeContracts.buildContractCall(multisig, "swapOwner",
+                [sentinelOwners, agentInstances[0].address, serviceOwnerAddress], nonce, 0, 0);
+            const signMessageData = await safeContracts.safeSignMessage(agentInstances[0], multisig, txHashData, 0);
+            await safeContracts.executeTx(multisig, txHashData, [signMessageData], 0);
```
Comment (kupermind) [inline test/ServiceRegistry.js:962]: Give up agent instance rights to the multisig to the service owner

File: test/ServiceRegistry.js [hunk 2]
```diff
@@ -905,6 +905,111 @@ describe("ServiceRegistry", function () {
...
+
+            // Change the existent multisig owners and threshold in a multisend transaction using the service owner access
+            let callData = [];
+            let txs = [];
+            nonce = await multisig.nonce();
+            // Add the addresses, but keep the threshold the same
+            for (let i = 0; i < agentInstances.length; i++) {
+                callData[i] = multisig.interface.encodeFunctionData("addOwnerWithThreshold", [agentInstances[i].address, 1]);
+                txs[i] = safeContracts.buildSafeTransaction({to: multisig.address, data: callData[i], nonce: 0});
+            }
```
Comment (kupermind) [inline test/ServiceRegistry.js:984]: Add new agent instances to the multisig

File: test/ServiceRegistry.js [hunk 3]
```diff
@@ -905,6 +905,111 @@ describe("ServiceRegistry", function () {
...
+            // Remove the original multisig owner and change the threshold
+            // Note that the prevOwner is the very first added address as it corresponds to the reverse order of added addresses
+            // The order in the gnosis safe multisig is as follows: sentinelOwners => agentInstances[last].address => ... =>
+            // => newOwnerAddresses[0].address => serviceOwnerAddress
+            callData.push(multisig.interface.encodeFunctionData("removeOwner", [agentInstances[0].address, serviceOwnerAddress,
+                newMaxThreshold]));
+            txs.push(safeContracts.buildSafeTransaction({to: multisig.address, data: callData[callData.length - 1], nonce: 0}));
```
Comment (kupermind) [inline test/ServiceRegistry.js:991]: Change the threshold and remove service owner from the multisig list of owners

File: test/ServiceRegistry.js [hunk 4]
```diff
@@ -905,6 +905,111 @@ describe("ServiceRegistry", function () {
...
+
+            let safeTx = safeContracts.buildMultiSendSafeTx(multiSend, txs, nonce);
+            await expect(
+                safeContracts.executeTxWithSigners(multisig, safeTx, [serviceOwner])
+            ).to.emit(multisig, "ExecutionSuccess");
+
+            // Pack the original multisig address
+            const data = ethers.utils.solidityPack(["address"], [multisig.address]);
+            // Redeploy the service using the newly updated multisig (same multisig address)
+            await serviceRegistry.connect(serviceManager).deploy(serviceOwnerAddress, serviceId,
```
Comment (kupermind) [inline test/ServiceRegistry.js:1006]: Redeploy service with the same multisig

**Review by mariapiamo — APPROVED** [review]

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 5]
```diff
@@ -31,40 +31,63 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
-    /// @param threshold Number of required confirmations for a multisig transaction.
-    /// @param data Packed data containing address of an existent gnosis safe multisig.
+    /// @param owners Set of updated multisig owners.
+    /// @param threshold Updated number for multisig transaction confirmations.
+    /// @param data Packed data containing address of an existent gnosis safe multisig and a payload to call the multisig with.
     /// @return multisig Address of a multisig (proxy).
     function create(
         address[] memory owners,
         uint256 threshold,
         bytes memory data
```
Comment (kupermind) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:56]: The data contains the proxy multisig address that was deployed, and the payload to call the `execTransaction()` on that multisig to modify it.

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 6]
```diff
@@ -31,40 +31,63 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
-        if (data.length != DEFAULT_DATA_LENGTH) {
+        uint256 dataLength = data.length;
+        if (dataLength < DEFAULT_DATA_LENGTH) {
             revert IncorrectDataLength(DEFAULT_DATA_LENGTH, data.length);
         }
 
-        // Read the proxy multisig address (20 bytes)
+        // Read the proxy multisig address (20 bytes) and the multisig-related data
         assembly {
             multisig := mload(add(data, DEFAULT_DATA_LENGTH))
```
Comment (kupermind) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:67]: Get the proxy multisig address

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 7]
```diff
@@ -31,40 +31,63 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
+        bytes memory payload;
+        if (dataLength > DEFAULT_DATA_LENGTH) {
+            uint256 payloadLength = dataLength - DEFAULT_DATA_LENGTH;
+            payload = new bytes(payloadLength);
+            for (uint256 i = 0; i < payloadLength; ++i) {
+                payload[i] = data[i + DEFAULT_DATA_LENGTH];
+            }
+
+            // Call the multisig with the provided payload
+            (bool success, ) = multisig.call(payload);
```
Comment (kupermind) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:80]: Get the multisig payload and execute it.

File: test/ServiceRegistry.js [hunk 8]
```diff
@@ -905,6 +911,230 @@ describe("ServiceRegistry", function () {
...
+                    { type: "uint256", name: "baseGas" },
+                    { type: "uint256", name: "gasPrice" },
+                    { type: "address", name: "gasToken" },
+                    { type: "address", name: "refundReceiver" },
+                    { type: "uint256", name: "nonce" },
+                ]
+            };
+
+            // Get the signature of a multisend transaction
+            const signatureBytes = await serviceOwner._signTypedData({verifyingContract: multisig.address, chainId: chainId}, EIP712_SAFE_TX_TYPE, safeTx);
```
Comment (kupermind) [inline test/ServiceRegistry.js:1119]: Form the multisend transaction for the multisig and sign it

File: test/ServiceRegistry.js [hunk 9]
```diff
@@ -905,6 +911,230 @@ describe("ServiceRegistry", function () {
...
+
+            // Forming Gnosis Safe transaction bytes data: the multisig address itself plus the execTransaction() data
+            // that forms the multisend data with the signature bytes to be executed by the multisig proxy
+            const safeExecData = gnosisSafe.interface.encodeFunctionData("execTransaction", [safeTx.to, safeTx.value,
+                safeTx.data, safeTx.operation, safeTx.safeTxGas, safeTx.baseGas, safeTx.gasPrice, safeTx.gasToken,
+                safeTx.refundReceiver, signatureBytes]);
```
Comment (kupermind) [inline test/ServiceRegistry.js:1125]: Build up the payload for the multisig exec function with all the required parameters and a signature

File: test/ServiceRegistry.js [hunk 10]
```diff
@@ -905,6 +911,230 @@ describe("ServiceRegistry", function () {
...
+
+            // Add the multisig address on top of the multisig exec transaction data
+            const packedData = ethers.utils.solidityPack(["address", "bytes"], [multisig.address, safeExecData]);
```
Comment (kupermind) [inline test/ServiceRegistry.js:1128]: Add a proxy multisig on top of the payload to be decomposed inside the `create()` function

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 11]
```diff
@@ -31,40 +31,63 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
 /// @author Aleksandr Kuperman - <aleksandr.kuperman@valory.xyz>
 contract GnosisSafeSameAddressMultisig {
     // Default data size to be parsed as an address of a Gnosis Safe multisig proxy address
+    // This exact size suggests that all the changes to the multisig have been performed and only validation is needed
     uint256 public constant DEFAULT_DATA_LENGTH = 20;
 
-    /// @dev Verifies the existent gnosis safe multisig for changed owners and threshold.
-    /// @notice The multisig is supposedly updated before reaching this step such that new multisig is not created.
+    /// @dev Updated and verifies the existent gnosis safe multisig for changed owners and threshold.
+    /// @notice The multisig could be updated before reaching this step such that new multisig is not created.
```
Comment (DavidMinarsch) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:46]: Not very clear

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 12]
```diff
@@ -31,40 +31,63 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
 
-        // Read the proxy multisig address (20 bytes)
+        // Read the proxy multisig address (20 bytes) and the multisig-related data
         assembly {
             multisig := mload(add(data, DEFAULT_DATA_LENGTH))
         }
 
-        // Get the provided multisig owners and threshold
+        // Read the payload, if any, that is going to change the multisig ownership and threshold
+        bytes memory payload;
```
Comment (DavidMinarsch) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:71]: move inside the if?

**Review by DavidMinarsch — COMMENTED** [review]
LGTM so far

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol (hunk 12 above)
Comment (kupermind) [reply contracts/multisigs/GnosisSafeSameAddressMultisig.sol:71]: sure :D

File: contracts/multisigs/GnosisSafeSameAddressMultisig.sol [hunk 13]
```diff
@@ -31,41 +31,70 @@ error WrongNumOwners(uint256 expected, uint256 provided);
...
+    /// @notice This function operates with existent multisig proxy that is requested to be updated in terms of
+    ///         the set of owners' addresses and the threshold. There are two scenarios possible:
+    ///         1. The multisig proxy is already updated before reaching this function. Then the multisig address
+    ///            must be passed as a payload such that its owners and threshold are verified against those specified
+    ///            in the argument list.
+    ///         2. The multisig proxy is not yet updated. Then the multisig address must be passed in a packed bytes of
+    ///            data along with the Gnosis Safe `execTransaction()` function arguments packed payload. That payload
+    ///            is going to modify the mulsisig proxy as per its signed transaction. At the end, the updated multisig
+    ///            proxy is going to be verified with the provided set of owners' addresses and the threshold.
+    ///         Note that the order of owners' addresses in the multisig is in reverse order, see comments below.
```
Comment (DavidMinarsch) [inline contracts/multisigs/GnosisSafeSameAddressMultisig.sol:55]: Please remove “see comments below” and replace with full explanation (so natspec is standalone).

**Review by DavidMinarsch — COMMENTED** [review]
Please add a negative test case, where the wrong agent instances and threshold are encoded in the calldata.

**Review by DavidMinarsch — APPROVED** [review]

File: test/ServiceRegistry.js [hunk 14]
```diff
@@ -1147,6 +1159,263 @@ describe("ServiceRegistry", function () {
...
+
+                // Try to deploy when passing the incorrect tx signature
+                safeExecErrData = gnosisSafe.interface.encodeFunctionData("execTransaction", [safeTx.to, safeTx.value,
+                    safeTx.data, safeTx.operation, safeTx.safeTxGas, safeTx.baseGas, safeTx.gasPrice, safeTx.gasToken,
+                    safeTx.refundReceiver, "0xabcd"]);
+                packedErrData = ethers.utils.solidityPack(["address", "bytes"], [multisig.address, safeExecErrData]);
+                await expect(
+                    serviceRegistry.connect(serviceManager).deploy(serviceOwnerAddress, serviceId,
+                        gnosisSafeSameAddressMultisig.address, packedErrData)
+                ).to.be.revertedWith("MultisigExecFailed");
```
Comment (kupermind) [inline test/ServiceRegistry.js:1408]: This two cases show that the safe transaction reverts with incorrectly constructed data fields.

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**